dvc/ignore.py

```diff
@@ -107,8 +107,8 @@ def matches(pattern, path, is_dir) -> bool:
 
         result = False
 
-        for ignore, pattern in self.ignore_spec[::-1]:
-            if matches(pattern, path, is_dir):
+        for ignore, patterns in self.ignore_spec[::-1]:
+            if any(matches(pattern, path, is_dir) for pattern in patterns):
```

Comment:
Hi, @hiroto7 I have a problem here, previously we concatenate these reg expressions mainly for the performance. And actually, in the very beginning, we use pathspaces's API to do these checks but it's really slow and the inner implementation is like this kind of nested for loop. Any better methods to solve this?

Reply:
@karajan1001  Oh, I didn't know that the regex concatenation is deliberate, and I have noticed you have previously mentioned a bottleneck in pathspec's API in https://github.com/iterative/dvc/issues/3869#issuecomment-635975677.

Although pathspec seems to try to check all regexes even if any pattern matches soonly, the `any()` function does short-circuit evaluation. So I think my PR never produces unnecessary regex matchings.

Reply:
Thank you for the benchmark result.

If that's the case, it seems that simply looping regexes one by one, as I did, is not a good idea.
Concatenating regexes is likely to be effective, unfortunately, but I have no idea about any other way.

Reply:
I had two suggestions here, 

1. use some kind of hacky way to override the regex pattern created by pathspec. (replace `<ps_d>` with `<ps_d1>', `<ps_d2>', ... `<ps_dn>' ) 
2. Another more radical choice is to contribute back to `pathspec` to improve its performance and then directly use the API of it.